Add initial support for Keyboard Lock

xingri opened the pull request:
#### ceba7343025b0e47c2903877198c8bdd5aeda851
<pre>
Add initial support for Keyboard Lock
<a href="https://bugs.webkit.org/show_bug.cgi?id=264843">https://bugs.webkit.org/show_bug.cgi?id=264843</a>

Reviewed by Brent Fulgham.

Implementation of Browser Keyboard Lock implementation as FullScreen API option.

Tests: fullscreen/full-screen-keyboard-lock-option-enabled.html
       fullscreen/full-screen-keyboard-lock-option.html
* LayoutTests/TestExpectations:
* LayoutTests/fullscreen/full-screen-keyboard-lock-option-enabled-expected.txt: Added.
* LayoutTests/fullscreen/full-screen-keyboard-lock-option-enabled.html: Added.
* LayoutTests/fullscreen/full-screen-keyboard-lock-option-expected.txt: Added.
* LayoutTests/fullscreen/full-screen-keyboard-lock-option.html: Added.
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/DocumentFullscreen.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::requestFullscreen):
* Source/WebCore/dom/FullscreenOptions.idl:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::EventHandler):
(WebCore::EventHandler::clear):
(WebCore::EventHandler::holdEscKeyEventTimerFired):
(WebCore::EventHandler::internalKeyEvent):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/Page.h:
(WebCore::Page::setHardwareKeyboardAttached):
(WebCore::Page::hardwareKeyboardAttached const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_toolbarsAreVisible):

Canonical link: <a href="https://commits.webkit.org/300356@main">https://commits.webkit.org/300356@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e3a0cd2c02e434524d2d98a73c593eeb4b3abd03

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122207 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41910 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32579 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128785 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74301 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124083 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42624 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50503 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92893 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61748 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125159 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34004 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109438 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73547 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33024 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27600 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72271 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103512 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27794 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131531 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/120733 "Build is in progress. Recent messages:") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49146 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37391 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101454 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49521 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105649 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101324 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25703 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46695 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24814 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45897 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49003 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54738 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48473 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51823 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50153 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fullscreen/full-screen-keyboard-lock-option-enabled.html:
```diff
@@ -0,0 +1,57 @@
+    <title>This tests that full screen mode with keyboard lock option allows ESC key event.</title>
+</head>
+<body>
+    <div style="container-type: size; width: 100%; height: 100%;" id="target">
```
Is there a reason to use `container-type: size` here?

xingri replied:
@nt1m  Thanks, fixed it.

nt1m commented on LayoutTests/fullscreen/full-screen-keyboard-lock-option-enabled.html:
```diff
@@ -0,0 +1,57 @@
+            callback = null;
+            testExpected("document.webkitIsFullScreen", true);
+            if (window.testRunner && eventSender.keyDown) {
+                waitForEvent(document, 'keypress', keyPressed);
```
nit: double quotes

nt1m commented on LayoutTests/fullscreen/full-screen-keyboard-lock-option-enabled.html:
```diff
@@ -0,0 +1,57 @@
+                callback = spanEnteredFullScreen;
+
+                waitForEvent(document, 'webkitfullscreenchange', fullscreenChanged);
+                addEventListener("fullscreenchange", resolve, { once: true });
```
Kind of strange that you're missing promises and event listeners here, can you only use one of them?

It doesn't seem like the promise is needed here.

xingri replied:
@nt1m  Thanks for review the codes. Will update the test cases.

nt1m commented on LayoutTests/fullscreen/full-screen-keyboard-lock-option.html:
Same comments apply to this file

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4434,9 +4434,19 @@ void Element::webkitRequestFullscreen()
+            promise->reject(ExceptionCode::NotSupportedError, "options.keyboardLock is unavailable."_s);
+            return;
+        }
+    }
```
```suggestion
void Element::requestFullscreen(FullscreenOptions&& options, RefPtr<DeferredPromise>&& promise)
{
    bool optionsEnabled = document().settings().fullScreenKeyboardLock();
    if (optionsEnabled)
        protectedDocument()->fullscreenManager().setKeyboardLockMode(options.keyboardLock);
    else
        ASSERT(options.keyboardLock == FullscreenOptions::KeyboardLock::None);
```

xingri replied:
@nt1m I do not think the ASSERT can generate promise rejection event on the non KeyboardLock mode.
So I would like to keep the current implementation AS-IS unless there is a way to auto generate the event through the ASSERT statement. How do you think?

marcoscaceres commented on LayoutTests/fullscreen/full-screen-keyboard-lock-option-enabled.html:
```diff
@@ -0,0 +1,52 @@
+
+        var spanEnteredFullScreen = function() {
+            callback = null;
+            testExpected("document.webkitIsFullScreen", true);
```
we should be testing against the real fullscreen API here, not the webkit one.

brentfulgham replied:
@xingri Can you please correct this? We want to make sure we use the modern API, not the old webkit-specific thing we want to do away with.

Or are you finding the specs-compliant API not to work here for some reason? If so, please file a bug so we can investigate.

marcoscaceres commented on LayoutTests/fullscreen/full-screen-keyboard-lock-option.html:
```diff
@@ -0,0 +1,52 @@
+
+        window.internals.settings.setFullScreenKeyboardLock(false);
+
+        var callback;
```
I think the tests could be simplified a lot... they seem to be using old JS idioms... I haven't tried this, but could you rewrite this test and the other more as:
```JS
window.internals.settings.setFullScreenKeyboardLock(true);
var event;
runWithKeyDown(async () => {
    try {
        const target = document.querySelector("#target");
        await target.requestFullscreen({ keyboardLock: "browser" });
        testExpected("document.fullscreenElement", true);
        event = await new Promise((resolve) => {
            waitForEvent(document, "keypress", resolve, /*endit=*/ false, /*once=*/ true);
            eventSender.keyDown("escape");
        });
        testExpected("event.keyCode", 27); // 'Escape'
    } catch (err) {
        logResult(
            false,
            `An error occurred while trying to enter full-screen mode: ${err.message}`
        );
    } finally {
        endTest();
    }
});
```
I think there is also a way of not having to check the global `var event` either...

brentfulgham replied:
@xingri Did you see this comment? Can you adopt @marcoscaceres 's suggestion, or help us understand why it doesn't work?

marcoscaceres commented on LayoutTests/fullscreen/full-screen-keyboard-lock-option-enabled.html:
```diff
@@ -0,0 +1,52 @@
+    <script src="full-screen-test.js"></script>
+    <script>
+
+        window.internals.settings.setFullScreenKeyboardLock(true);
```
This seems to be using old JS idioms and legacy fullscreen events. I haven't tried the below, but could you maybe simplify the tests by rewriting it in this style: 

```JS
window.internals.settings.setFullScreenKeyboardLock(true);
var event;
runWithKeyDown(async () => {
    const target = document.querySelector("#target");
    try {
        await target.requestFullscreen({ keyboardLock: "browser" });
        testExpected("document.fullscreenElement", true);
        event = await new Promise((resolve) => {
            waitForEvent(
                document,
                "keypress",
                resolve,
                /*endit=*/ false,
                /*once=*/ true
            );
            eventSender.keyDown("escape");
        });
        testExpected("event.key", "Escape");
    } catch (err) {
        logResult(
            false,
            `An error occurred while trying to enter full-screen mode: ${err.message}`
        );
    } finally {
        endTest();
    }
});
```

marcoscaceres replied:
I think there should even be some way of getting rid of that global `var event` and testing for the key code without relying on the global scope.

xingri replied:
@marcoscaceres  Thanks for the advice. Updated the test cases.

brentfulgham commented on Source/WebCore/dom/FullscreenManager.h:
```diff
@@ -135,6 +140,9 @@ class FullscreenManager final : public CanMakeWeakPtr<FullscreenManager>, public
     bool m_isAnimatingFullscreen { false };
 
+    // Fullscreen Keyboard Lock
+    FullscreenOptions::KeyboardLock m_keyboardLockMode = FullscreenOptions::KeyboardLock::None;
```
Nit: We should use `{ FullscreenOptions::KeyboardLock::None };` syntax here.

brentfulgham commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -3767,8 +3781,17 @@ bool EventHandler::internalKeyEvent(const PlatformKeyboardEvent& initialKeyEvent
+            if (fullscreenManager->isBrowserKeyboardLockEnabled()) {
+                // check 1.5 ms ESC hold to cancel Full Screen
+                if (!m_holdEscKeyEventTimer.isActive())
+                    m_holdEscKeyEventTimer.startOneShot(1500_ms);
```
The [draft specification](https://wicg.github.io/keyboard-lock/#security) suggests "longer than 2 seconds", which seems quite long. Your 1.5 second choice is preferable!

brentfulgham commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4601,9 +4601,19 @@ void Element::webkitRequestFullscreen()
+// FIXME: Only KeyboardLock option is currently considered.
+void Element::requestFullscreen(FullscreenOptions&& options, RefPtr<DeferredPromise>&& promise)
 {
+    bool optionsEnabled = document().settings().fullScreenKeyboardLock();
```
I think we should restrict this to cases where `PAL::currentUserInterfaceIdiomIsDesktop()` is true, like we do for Pointer Lock. Are mobile devices like iPhones an important use case for keyboard locking like this? I sort of doubt it unless users regularly attach bluetooth keyboard or something.

JeffreyBlanz replied:
Mobile devices like iPhone, iPad, and the Apple Vision Pro (so all of iOS) would definitely use Keyboard Lock API as well.

It is used when you want to use keyboard + mouse input for gaming on those devices. The keyboard lock is needed to use the escape key to open menus on those mobile devices as well.

xingri replied:
@JeffreyBlanz  I agree with you that mobile and Vision Pro are important use cases for Keyboard Lock, especially for gaming scenarios. However, to reduce complexity and risk for the initial implementation, as @brentfulgham  suggested we start by enabling this feature only when PAL::currentUserInterfaceIdiomIsDesktop() is true (similar to Pointer Lock). Once we validate the desktop experience, I hope we can extend support to iOS and other platforms in a follow-up.

aprotyas replied:
Probably valid to start with `currentUserInterfaceIdiomIsDesktop() | currentUserInterfaceIdiomIsVision()`, given the visionOS platform is desktop-class, too.

xingri replied:
@aprotyas  Thanks for the information.

brentfulgham replied:
@aprotyas I kind of think we should change `currentUserInterfaceIdiomIsDesktop()` to include Vision. What do you think? Or should it be a separate patch?

brentfulgham commented on LayoutTests/fullscreen/full-screen-keyboard-lock-option-enabled-expected.txt:
```diff
@@ -0,0 +1,7 @@
+Test passes if ESC key received by the test case after enter to fullscreen.
```
Is it possible for these to be tested as WPT instead of in our custom testing infrastructure? Or does this test things that are unique to WebKit ports and would not be appropriate for WPT itself?

xingri replied:
@brentfulgham  I agree that most of this can (and should) be covered in Web Platform Tests (WPT) since we’re validating the proposed Keyboard Lock API for the Web. WPT already has a keyboard-lock/ area and IDL coverage we can extend, and we can use testdriver.js for user activation and key input instead of custom infrastructure.
We can land these as tentative tests while the spec discussion continues. That said, I think it’s valuable to get the initial patch landed in WebKit first so we can validate the implementation and gather feedback before finalizing the WPT coverage.

brentfulgham replied:
FYI: I don't propose you move these to WPT in this PR. But maybe as a follow-up we could do that, to help enforce our expected behavior in Chrome and elsewhere.

brentfulgham commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4601,9 +4601,19 @@ void Element::webkitRequestFullscreen()
 {
+    bool optionsEnabled = document().settings().fullScreenKeyboardLock();
+    if (optionsEnabled)
+        protectedDocument()->fullscreenManager().setKeyboardLockMode(options.keyboardLock);
```
@xingri I think this would be better if we could restrict keyboard lock behavior to cases where the user has a physical keyboard (so always available on macOS, available with hardware keyboard on iOS/Vision OS). I can imagine sites accidentally getting this activated for users with a software keyboard. How do they long-press escape to exit in that case? I think the keyboard goes away if there is not active text input field? (I might be mistaken about that).

brentfulgham replied:
Maybe something using 
`WebProcessPool::cachedHardwareKeyboardState().isAttached == true`

xingri replied:
@brentfulgham The idea of limiting the Keyboard Lock API on hardware keyboards strikes me as reasonable.

brentfulgham reviewed (changes requested):
I think this patch is a great start! I think we need a few things to make this ready to land:
1. We should limit the ability to enable Full Screen Keyboard Lock to cases where a hardware keyboard is present.
2. We should limit this feature to cases where the `PAL::currentUserInterfaceIdiomIsDesktop()` is true.

In follow-up patches, we should make Vision Pro part of Desktop (not needed in this patch) and perhaps push these tests to WPT.

brentfulgham commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5052,9 +5056,23 @@ void Element::webkitRequestFullscreen()
+#else
+    bool optionsEnabled = document().settings().fullScreenKeyboardLock();
+#endif
+    if (optionsEnabled && document().page() && document().page()->hardwareKeyboardAttached())
```
Very nice!

xingri replied:
updated.

brentfulgham commented on Source/WebCore/dom/FullscreenOptions.idl:
```diff
@@ -32,7 +32,7 @@
 
 [
-    Conditional=FULLSCREEN_API, EnabledBySetting=FullScreenKeyboardLock
+    Conditional=FULLSCREEN_API
```
Why are we removing the `EnabledBySetting` here? I think we would want this to be controllable by the User Agent in Page Settings, so we could allow keyboard lock on gaming sites, but prevent general sites without this need from capturing the keyboard.

I propose we leave this in place for this patch, and discuss moving the setting to `stable` in a follow-up patch. It's already marked as `testable` so your test cases should be fine even without this bit of your change.

xingri replied:
@brentfulgham  Agreed on your review. Previously, I just did not want to the default test case ended up timed out. 
But that's only the concern I had and I thought this might be better to leave to be timed out in case of default test.

brentfulgham reviewed (changes requested):
I think this looks great! Could you answer the two questions about test cases, and revert the bit where you remove the `FullScreenKeyboardLock` protection in the IDL? Then I think this is ready to land.

brentfulgham reviewed (approved):
Thank you for restoring the feature flag (for now). r=me

xingri committed:
Add initial support for Keyboard Lock
https://bugs.webkit.org/show_bug.cgi?id=264843

Reviewed by Brent Fulgham.

Implementation of Browser Keyboard Lock implementation as FullScreen API option.

Tests: fullscreen/full-screen-keyboard-lock-option-enabled.html
       fullscreen/full-screen-keyboard-lock-option.html
* LayoutTests/TestExpectations:
* LayoutTests/fullscreen/full-screen-keyboard-lock-option-enabled-expected.txt: Added.
* LayoutTests/fullscreen/full-screen-keyboard-lock-option-enabled.html: Added.
* LayoutTests/fullscreen/full-screen-keyboard-lock-option-expected.txt: Added.
* LayoutTests/fullscreen/full-screen-keyboard-lock-option.html: Added.
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/DocumentFullscreen.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::requestFullscreen):
* Source/WebCore/dom/FullscreenOptions.idl:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::EventHandler):
(WebCore::EventHandler::clear):
(WebCore::EventHandler::holdEscKeyEventTimerFired):
(WebCore::EventHandler::internalKeyEvent):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/Page.h:
(WebCore::Page::setHardwareKeyboardAttached):
(WebCore::Page::hardwareKeyboardAttached const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_toolbarsAreVisible):

Canonical link: https://commits.webkit.org/300356@main
